### PR TITLE
[LLVM][SelectionDAG] Ensure Constant[FP]SDnode only store references to scalar Constant{Int,FP}.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
This fixes a failure path when the use-constant-##-for-###-splat IR options are enabled.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 1]
```diff
@@ -1656,6 +1656,11 @@ SDValue SelectionDAG::getConstant(const ConstantInt &Val, const SDLoc &DL,
   EVT EltVT = VT.getScalarType();
   const ConstantInt *Elt = &Val;
 
+  // Vector splats are explicit within the DAG, with ConstantSDNode holding the
+  // to be splatted scalar ConstantInt.
+  if (isa<VectorType>(Elt->getType()))
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:1661]: Check VT.isVector instead? I assume these have to match?

Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:1661]: They don't have to match. In most cases `Val` will be the scalar you want to splat.  The new use case is when SelectionDAGBuilder passes in the new style ConstantInt which can have a vector result type when `use-constant-int-for-scalable-vector-splat` is enabled.

File: llvm/include/llvm/CodeGen/SelectionDAGNodes.h [hunk 2]
```diff
@@ -1678,6 +1678,7 @@ class ConstantSDNode : public SDNode {
       : SDNode(isTarget ? ISD::TargetConstant : ISD::Constant, 0, DebugLoc(),
                VTs),
         Value(val) {
+    assert(!isa<VectorType>(val->getType()) && "Unexpected vector type!");
```
Comment (david-arm) [inline llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1681]: This is just a thought, but is it worth tying these asserts down even further, i.e.

  assert(val->getType()->isIntegerTy() && "Unexpected type!");

and similarly for the FP node? At the moment, these nodes still won't assert for any other unusual types.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 3]
```diff
@@ -1656,6 +1656,11 @@ SDValue SelectionDAG::getConstant(const ConstantInt &Val, const SDLoc &DL,
   EVT EltVT = VT.getScalarType();
   const ConstantInt *Elt = &Val;
 
+  // Vector splats are explicit within the DAG, with ConstantSDNode holding the
+  // to be splatted scalar ConstantInt.
```
Comment (david-arm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:1660]: nit: to-be-splatted?

File: llvm/include/llvm/CodeGen/SelectionDAGNodes.h (hunk 2 above)
Comment (paulwalker-arm) [reply llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1681]: That's handled as part of `ConstantInt` construction and thus here we can safely assume `val` has a valid type for its class.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp (hunk 3 above)
Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: Done.

**Review by david-arm — APPROVED** [review]
LGTM. Beautiful, just beautiful!

**Pull request merged**